File: packages/guards/src/agent-goal-drift.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface AgentGoalDriftOptions {
  action: 'block' | 'warn';
  /** Keywords that define the agent's goal */
  goalKeywords: string[];
  /** Minimum ratio of keywords present (default 0.3) */
  driftThreshold?: number;
}

export function agentGoalDrift(options: AgentGoalDriftOptions): Guard {
  const threshold = options.driftThreshold ?? 0.3;
  const keywords = options.goalKeywords.map((k) => k.toLowerCase());

  return {
    name: 'agent-goal-drift',
    version: '0.1.0',
    description: 'Detects when an agent drifts from its stated goal',
    category: 'security',
    supportedStages: ['input', 'output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const lower = text.toLowerCase();
      let found = 0;

      for (const kw of keywords) {
        if (lower.includes(kw)) found++;
      }

      const ratio = keywords.length > 0 ? found / keywords.length : 1;
      const triggered = ratio < threshold;

      return {
        guardName: 'agent-goal-drift',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        latencyMs: Math.round(performance.now() - start),
        details: triggered
          ? { keywordRatio: ratio, threshold, foundKeywords: found, totalKeywords: keywords.length }
          : undefined,
      };
    },
  };
}

File: python/open_guardrail/guards/agent_goal_drift.py
"""Detects when an agent drifts from its stated goal."""
from __future__ import annotations

import time
from typing import List

from open_guardrail.core import GuardResult


class _AgentGoalDrift:
    def __init__(
        self,
        *,
        action: str = "block",
        goal_keywords: List[str],
        drift_threshold: float = 0.3,
    ) -> None:
        self.name = "agent-goal-drift"
        self.action = action
        self.keywords = [k.lower() for k in goal_keywords]
        self.threshold = drift_threshold

    def check(self, text: str, stage: str = "input") -> GuardResult:
        start = time.perf_counter()
        lower = text.lower()
        found = sum(1 for kw in self.keywords if kw in lower)
        ratio = found / len(self.keywords) if self.keywords else 1.0
        triggered = ratio < self.threshold
        elapsed = (time.perf_counter() - start) * 1000
        return GuardResult(
            guard_name="agent-goal-drift",
            passed=not triggered,
            action=self.action if triggered else "allow",
            message=f"Goal drift detected: only {found}/{len(self.keywords)} keywords present ({ratio:.0%})" if triggered else None,
            latency_ms=round(elapsed, 2),
            details={"keyword_ratio": ratio, "threshold": self.threshold, "found_keywords": found, "total_keywords": len(self.keywords)} if triggered else None,
        )


def agent_goal_drift(
    *,
    action: str = "block",
    goal_keywords: List[str],
    drift_threshold: float = 0.3,
) -> _AgentGoalDrift:
    return _AgentGoalDrift(
        action=action,
        goal_keywords=goal_keywords,
        drift_threshold=drift_threshold,
    )
